### PR TITLE
dockerhub-images: use ctr instead of docker CLI

### PR DESCRIPTION
**Pull request by egernst**
Use containerd CLI instead of docker for 2.0 testing of our kata v2 shims.

### CODE REVIEW

**Comment by GabyCT** [comment]
@egernst where this test will be triggered?

**Comment by chavafg** [comment]
@egernst I still see a `docker rmi` in the `teardown`

File: integration/popular_docker_hub_images/popular_docker_images.bats [hunk 1]
```diff
@@ -21,475 +21,635 @@ setup() {
 	clean_env
```
Comment (GabyCT) [inline integration/popular_docker_hub_images/popular_docker_images.bats:21]: @egernst this function is using docker
https://github.com/kata-containers/tests/blob/2.0-dev/lib/common.bash#L221

Comment (egernst) [reply integration/popular_docker_hub_images/popular_docker_images.bats]: addressed.

File: integration/popular_docker_hub_images/popular_docker_images.bats [hunk 2]
```diff
@@ -11,490 +11,560 @@ source ${BATS_TEST_DIRNAME}/../../lib/common.bash
...
+	image="docker.io/library/zookeeper:latest"
+	sudo ctr image pull $image
+	sudo ctr run --rm --runtime=$RUNTIME -t $image foobar zkServer.sh
 }
 
 teardown() {
-	clean_env
-	docker rmi $image
+	clean_env_ctr
+        sudo ctr image rm $image
```
Comment (GabyCT) [inline integration/popular_docker_hub_images/popular_docker_images.bats:567]: extra space (not aligned)

**Comment by egernst** [comment]
@GabyCT updated, PTAL.

I can test this locally. I saw a couple of tests failed, but it completed and cleaned up after itself. I think the failures may just be local networking issues.

File: integration/popular_docker_hub_images/popular_docker_images.bats [hunk 3]
```diff
@@ -11,490 +11,560 @@ source ${BATS_TEST_DIRNAME}/../../lib/common.bash
...
+	sudo ctr run --rm --runtime=$RUNTIME -t $image foobar zkServer.sh
 }
 
 teardown() {
-	clean_env
-	docker rmi $image
+	clean_env_ctr
+	sudo ctr image rm $image
 	# Check that processes are not running
 	run check_processes
```
Comment (GabyCT) [inline integration/popular_docker_hub_images/popular_docker_images.bats:569]: @egernst this function is using docker too

Comment (egernst) [reply integration/popular_docker_hub_images/popular_docker_images.bats:569]: that one is a pain. I'm not sure how much utility there

Comment (egernst) [reply integration/popular_docker_hub_images/popular_docker_images.bats:569]: Perhaps we can *just* check for containerd-shim-kata-v2 in our case? ie

pgrep -f `which containerd-shim-kata-v2` ?

Comment (egernst) [reply integration/popular_docker_hub_images/popular_docker_images.bats]: Horrible bash applied. PTAL

Comment (fidencio) [reply integration/popular_docker_hub_images/popular_docker_images.bats]: @egernst, switching the check in the pgrep is fine as we only care for `containerd-shim-kata-v2`.
About not having kata-runtime CLI, we still do build and ship kata-runtime in 2.0 and kata-env / kata-check should work (I just tested it locally).

File: lib/common.bash [hunk 4]
```diff
@@ -158,6 +158,20 @@ extract_kata_env(){
 	NETMON_PATH="/usr/libexec/kata-containers/kata-netmon"
 }
 
+# Checks that containerd-shim-kata-v2 processes are not running
+check_ctr_shim_processes() {
```
Comment (fidencio) [inline lib/common.bash:162]: nitpick:
We could still call `kata-runtime kata-env` without any issue and it'd bring some info about the system.
But I totally understand the reason for not calling it.

File: integration/popular_docker_hub_images/popular_docker_images.bats (hunk 3 above)
Comment (fidencio) [reply integration/popular_docker_hub_images/popular_docker_images.bats]: That bash script is actually not as bad as you painted it. :-)
LGTM!

File: lib/common.bash (hunk 4 above)
Comment (egernst) [reply lib/common.bash:162]: That ends up doing a lot of parsing, etc, to get the same info I think we can get from which.

Long term that may be more robust, but I probably won't have cycles to add that.

**Comment by GabyCT** [comment]
can we test this PR? we can triggered the jenkins job that @chavafg created to test this PR and to see that everything is working before merging in case that further changes need to be applied

**Review by jodh-intel — APPROVED** [review]
Thanks @egernst !

**Comment by egernst** [comment]
There are many tests which end up failing. I think some of it has to do with a lack of networking when using ctr, perhaps different container images being utilized, and syntax failures for the cmdline given to the container.

I don't have a lot of cycles to spend digging through each one of them, but they are failing on runc as well as containerd-shim-kata-v2.

**Comment by jodh-intel** [comment]
@egernst - the errors seem to be mostly crashes in containerd, such as:

```
01:07:04 not ok 1 [display text] hello world in an alpine container
01:07:04 # (in test file popular_docker_images.bats, line 27)
01:07:04 #   `sudo ctr run --rm --runtime=$RUNTIME -t $image foobar sh -c "echo 'Hello, World':latest"' failed with status 2

    :

01:07:04 # unpacking linux/amd64 sha256:185518070891758909c9f839cf4ca393ee977ac378609f700f60a771a2dfe321...
01:07:04 # done
01:07:04 # panic: provided file is not a console
01:07:04 # 
01:07:04 # goroutine 1 [running]:
01:07:04 # github.com/containerd/containerd/vendor/github.com/containerd/console.Current(...)
01:07:04 # 	/tmp/cri-install-containerd.cOga/src/github.com/containerd/containerd/vendor/github.com/containerd/console/console.go:67
01:07:04 # github.com/containerd/containerd/cmd/ctr/commands/run.glob..func1(0xc000135760, 0x0, 0x0)
01:07:04 # 	/tmp/cri-install-containerd.cOga/src/github.com/containerd/containerd/cmd/ctr/commands/run/run.go:164 +0xffd
01:07:04 # github.com/containerd/containerd/vendor/github.com/urfave/cli.HandleAction(0x55d8cbf8e180, 0x55d8cc0b39f8, 0xc000135760, 0xc000135760, 0x0)
01:07:04 # 	/tmp/cri-install-containerd.cOga/src/github.com/containerd/containerd/vendor/github.com/urfave/cli/app.go:523 +0x11c
01:07:04 # github.com/containerd/containerd/vendor/github.com/urfave/cli.Command.Run(0x55d8cbbeae7c, 0x3, 0x0, 0x0, 0x0, 0x0, 0x0, 0x55d8cbbf5ad2, 0xf, 0x0, ...)
01:07:04 # 	/tmp/cri-install-containerd.cOga/src/github.com/containerd/containerd/vendor/github.com/urfave/cli/command.go:174 +0x570
01:07:04 # github.com/containerd/containerd/vendor/github.com/urfave/cli.(*App).Run(0xc00037c540, 0xc0000320a0, 0xa, 0xa, 0x0, 0x0)
01:07:04 # 	/tmp/cri-install-containerd.cOga/src/github.com/containerd/containerd/vendor/github.com/urfave/cli/app.go:276 +0x7b0
01:07:04 # main.main()
01:07:04 # 	github.com/containerd/containerd/cmd/ctr/main.go:37 +0x134
01:07:04 # docker.io/library/alpine:latest
```

If you look at the `procenv` build artifact, it shows:

```yaml
file descriptors:
  0:
    terminal: no
    valid: yes
    device: pipe:[124382]
  1:
    terminal: no
    valid: yes
    device: pipe:[124383]
  2:
    terminal: no
    valid: yes
    device: pipe:[124383]
```

So could you try running `ctr` without the `-t` as seeing if that helps?

**Comment by fidencio** [comment]
I'm really sorry for the disruption but in order to do the branch name changes announced [here](http://lists.katacontainers.io/pipermail/kata-dev/2021-January/001695.html) I'm adding the "do-not-merge" and the "needs-branch-adjustment" labels to the PR.

I'll add a new comment once the branch changes are done and once it happens I'd like to ask you to resubmit the PR against the new `main` branch, possibly linking to this PR in case you find the conversations / reviews already made important to proceed.

**Comment by fidencio** [comment]
Kata Containers repos are back to business.  Thanks a lot for your patience and I'm very happy to announce that there's no changes needed from your side as renaming the branches automagically reflected here.

If you notice something weird, some missing CI, please, do not hesitate on contacting us here or on #kata-dev on bit.ly/katacontainersslack.

Also, this is a good opportunity to have your PR rebased atop of `main`, to ensure your PR follow all the guidelines (https://github.com/kata-containers/community/blob/master/CONTRIBUTING.md#patch-format) enforced by this community, or even to re-evaluate whether the PR is currently needed.

Last but not least,  as the default branch has been renamed to `main`, if you have a local clone you'd need to update it and you do so by running:
```
git branch -m 2.0-dev main
git fetch origin
git branch -u origin/main main
```

Happy hacking!

**Comment by ariel-adam** [comment]
@egernst can we close this PR due to https://github.com/kata-containers/tests/pull/3540 ?

**Comment by ariel-adam** [comment]
@egernst closing this PR for now due to #3540.
Please reopen if needed.

**Pull request closed without merging**